File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/account/TransferActivity.java
package com.blogspot.e_kanivets.moneytracker.activity.account;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ArrayAdapter;

import androidx.annotation.Nullable;

import com.blogspot.e_kanivets.moneytracker.R;
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseBackActivity;
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController;
import com.blogspot.e_kanivets.moneytracker.controller.data.TransferController;
import com.blogspot.e_kanivets.moneytracker.databinding.ActivityTransferBinding;
import com.blogspot.e_kanivets.moneytracker.entity.data.Account;
import com.blogspot.e_kanivets.moneytracker.entity.data.Transfer;
import com.blogspot.e_kanivets.moneytracker.util.CrashlyticsProxy;
import com.blogspot.e_kanivets.moneytracker.util.validator.IValidator;
import com.blogspot.e_kanivets.moneytracker.util.validator.TransferValidator;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class TransferActivity extends BaseBackActivity {
    @SuppressWarnings("unused")
    private static final String TAG = "TransferActivity";

    @Inject
    TransferController transferController;
    @Inject
    AccountController accountController;

    private IValidator<Transfer> transferValidator;

    private List<Account> accountList;

    private ActivityTransferBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityTransferBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        initData();
        initToolbar();
        initViews();
    }

    private boolean initData() {
        getAppComponent().inject(TransferActivity.this);
        accountList = accountController.readActiveAccounts();
        return true;
    }

    private void initViews() {
        List<String> accounts = new ArrayList<>();
        for (Account account : accountList) {
            accounts.add(account.getTitle());
        }

        transferValidator = new TransferValidator(TransferActivity.this, binding);

        if (accounts.size() == 0) {
            accounts.add(getString(R.string.none));
            binding.spinnerFrom.setEnabled(false);
            binding.spinnerTo.setEnabled(false);
        }

        binding.spinnerFrom.setAdapter(new ArrayAdapter<>(TransferActivity.this,
                R.layout.view_spinner_item, accounts));

        binding.spinnerTo.setAdapter(new ArrayAdapter<>(TransferActivity.this,
                R.layout.view_spinner_item, accounts));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_transfer, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_done) {
            tryTransfer();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void tryTransfer() {
        CrashlyticsProxy.get().logButton("Done Transfer");
        if (doTransfer()) {
            CrashlyticsProxy.get().logEvent("Done Transfer");
            setResult(RESULT_OK);
            finish();
        }
    }

    @SuppressWarnings("SimplifiableIfStatement")
    private boolean doTransfer() {
        if (transferValidator.validate()) {
            Account fromAccount = accountList.get(binding.spinnerFrom.getSelectedItemPosition());
            Account toAccount = accountList.get(binding.spinnerTo.getSelectedItemPosition());
            double fromAmount = Double.parseDouble(binding.etFromAmount.getText().toString());
            double toAmount = Double.parseDouble(binding.etToAmount.getText().toString());

            return transferController.create(new Transfer(System.currentTimeMillis(),
                    fromAccount.getId(), toAccount.getId(), fromAmount, toAmount)) != null;
        } else {
            return false;
        }
    }
}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/account/TransferActivity.kt
package com.blogspot.e_kanivets.moneytracker.activity.account

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.ArrayAdapter
import com.blogspot.e_kanivets.moneytracker.R
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseBackActivity
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController
import com.blogspot.e_kanivets.moneytracker.controller.data.TransferController
import com.blogspot.e_kanivets.moneytracker.databinding.ActivityTransferBinding
import com.blogspot.e_kanivets.moneytracker.entity.data.Account
import com.blogspot.e_kanivets.moneytracker.entity.data.Transfer
import com.blogspot.e_kanivets.moneytracker.util.CrashlyticsProxy
import com.blogspot.e_kanivets.moneytracker.util.validator.IValidator
import com.blogspot.e_kanivets.moneytracker.util.validator.TransferValidator
import javax.inject.Inject

class TransferActivity : BaseBackActivity() {

    @Inject
    lateinit var transferController: TransferController

    @Inject
    lateinit var accountController: AccountController

    private lateinit var transferValidator: IValidator<Transfer>

    private lateinit var accountList: List<Account>

    private lateinit var binding: ActivityTransferBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityTransferBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initData()
        initToolbar()
        initViews()
    }

    private fun initData(): Boolean {
        appComponent.inject(this)
        accountList = accountController.readActiveAccounts()
        return true
    }

    private fun initViews() {
        val accounts = mutableListOf<String>()
        for (account in accountList) {
            accounts.add(account.title)
        }

        transferValidator = TransferValidator(this, binding)

        if (accounts.isEmpty()) {
            accounts.add(getString(R.string.none))
            binding.spinnerFrom.isEnabled = false
            binding.spinnerTo.isEnabled = false
        }

        binding.spinnerFrom.adapter = ArrayAdapter(this, R.layout.view_spinner_item, accounts)
        binding.spinnerTo.adapter = ArrayAdapter(this, R.layout.view_spinner_item, accounts)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_transfer, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_done -> {
            tryTransfer()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    private fun tryTransfer() {
        CrashlyticsProxy.instance.logButton("Try Transfer")
        if (doTransfer()) {
            CrashlyticsProxy.instance.logEvent("Done Transfer")
            setResult(RESULT_OK)
            finish()
        }
    }

    @SuppressWarnings("SimplifiableIfStatement")
    private fun doTransfer() = if (transferValidator.validate()) {
        val fromAccount = accountList[binding.spinnerFrom.selectedItemPosition]
        val toAccount = accountList[binding.spinnerTo.selectedItemPosition]
        val fromAmount = binding.etFromAmount.text.toString().toDouble()
        val toAmount = binding.etToAmount.text.toString().toDouble()

        transferController.create(
            Transfer(
                System.currentTimeMillis(),
                fromAccount.id,
                toAccount.id,
                fromAmount,
                toAmount
            )
        ) != null
    } else {
        false
    }
}
